File: Books/A_Comprehensive_Approach_to_ML_and_AI/matlab/decision_tree_tutorial.m
% This tutorial demonstrates a decision tree classifier built entirely from
% scratch. We generate a 2D dataset with two classes, recursively build a tree
% using the Gini impurity criterion, and visualize the decision boundary.

close all; clear; clc;
rng(1);  % For reproducibility

%% 1. Generate Synthetic Data
N = 100;
% Class +1: centered at (2,2)
X1 = randn(N,2) + 2;
% Class -1: centered at (-2,-2)
X2 = randn(N,2) - 2;
X = [X1; X2];
y = [ones(N,1); -ones(N,1)];  % Labels: +1 and -1

%% 2. Build Decision Tree from Scratch
maxDepth = 3;  % Limit tree depth for simplicity (similar to MaxNumSplits = 5)
tree = buildTree(X, y, 0, maxDepth);

%% 3. Visualize the Decision Boundary
% Create a grid over the feature space.
x_min = min(X(:,1)) - 1; x_max = max(X(:,1)) + 1;
y_min = min(X(:,2)) - 1; y_max = max(X(:,2)) + 1;
[xGrid, yGrid] = meshgrid(linspace(x_min, x_max, 200), linspace(y_min, y_max, 200));
gridPoints = [xGrid(:), yGrid(:)];

% Predict using the tree
preds = predictTree(tree, gridPoints);
preds = reshape(preds, size(xGrid));

% Plot data and decision regions.
figure;
hold on;
scatter(X(y==1,1), X(y==1,2), 50, 'bo', 'DisplayName', 'Class +1');
scatter(X(y==-1,1), X(y==-1,2), 50, 'bo', 'DisplayName', 'Class -1');
title('Decision Tree Classification');
xlabel('Feature 1'); ylabel('Feature 2');
legend('Clusters','Location','SE');
grid on;
hold off;

% save_all_figs_OPTION('results/decision_tree1','png',1)

% Plot data and decision regions.
figure;
hold on;
scatter(X(y==1,1), X(y==1,2), 50, 'bo', 'DisplayName', 'Class +1');
scatter(X(y==-1,1), X(y==-1,2), 50, 'ro', 'DisplayName', 'Class -1');
contourf(xGrid, yGrid, preds, [-1 0 1], 'LineWidth', 0.8, 'FaceAlpha', 0.1); % Decision boundary
colormap([1 0.8 0.8; 0.8 0.8 1]);  % Red-ish for class -1, blue-ish for class +1
title('Decision Tree Classification - Decision Boundary');
xlabel('Feature 1'); ylabel('Feature 2');
legend('Class 0','Class 1','Location','SE');
grid on;
hold off;

% save_all_figs_OPTION('results/decision_tree2','png',1)

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%                   Subfunctions for Decision Tree
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function tree = buildTree(X, y, depth, maxDepth)
    % buildTree recursively builds a decision tree classifier.
    %
    % Inputs:
    %   X       - n x d data matrix.
    %   y       - n x 1 vector of labels (+1 or -1).
    %   depth   - current depth of the tree.
    %   maxDepth- maximum allowed depth.
    %
    % Output:
    %   tree    - a struct representing the decision tree.
    %
    % The tree struct has the fields:
    %   .isLeaf      - boolean flag indicating if the node is a leaf.
    %   .prediction  - predicted class if node is a leaf.
    %   .feature     - feature index for splitting.
    %   .threshold   - threshold value for the split.
    %   .left, .right- subtrees (if not a leaf).
    
    % Stopping criteria: all labels identical, max depth reached, or too few samples.
    if (all(y == y(1))) || (depth >= maxDepth) || (size(X,1) < 2)
        tree.isLeaf = true;
        tree.prediction = majorityVote(y);
        return;
    end
    
    % Find best split using Gini impurity.
    [bestFeature, bestThreshold, bestImpurity, bestSplit] = findBestSplit(X, y);
    
    % If no valid split is found, create a leaf.
    if isempty(bestFeature)
        tree.isLeaf = true;
        tree.prediction = majorityVote(y);
        return;
    end
    
    % Create internal node.
    tree.isLeaf = false;
    tree.feature = bestFeature;
    tree.threshold = bestThreshold;
    
    % Partition data based on the best split.
    leftIndices = bestSplit;
    rightIndices = ~bestSplit;
    
    % If one partition is empty, make this node a leaf.
    if isempty(find(leftIndices,1)) || isempty(find(rightIndices,1))
        tree.isLeaf = true;
        tree.prediction = majorityVote(y);
        return;
    end
    
    tree.left = buildTree(X(leftIndices, :), y(leftIndices), depth+1, maxDepth);
    tree.right = buildTree(X(rightIndices, :), y(rightIndices), depth+1, maxDepth);
end

function [bestFeature, bestThreshold, bestImpurity, bestSplit] = findBestSplit(X, y)
    % findBestSplit searches over features and thresholds to find the split
    % that minimizes the weighted Gini impurity.
    %
    % Outputs:
    %   bestFeature   - index of the best feature.
    %   bestThreshold - threshold value for the best split.
    %   bestImpurity  - weighted impurity of the best split.
    %   bestSplit     - logical vector indicating the left split (X(:,feature) < threshold).
    
    [n, d] = size(X);
    bestImpurity = Inf;
    bestFeature = [];
    bestThreshold = [];
    bestSplit = [];
    
    for j = 1:d
        values = unique(X(:,j));
        for i = 1:length(values)
            threshold = values(i);
            split = X(:,j) < threshold;
            if sum(split) == 0 || sum(~split) == 0
                continue;  % Skip invalid splits.
            end
            impurityLeft = giniImpurity(y(split));
            impurityRight = giniImpurity(y(~split));
            weightedImpurity = (sum(split)/n)*impurityLeft + (sum(~split)/n)*impurityRight;
            if weightedImpurity < bestImpurity
                bestImpurity = weightedImpurity;
                bestFeature = j;
                bestThreshold = threshold;
                bestSplit = split;
            end
        end
    end
end

function gini = giniImpurity(y)
    % giniImpurity computes the Gini impurity for binary labels.
    n = length(y);
    if n == 0
        gini = 0;
    else
        p = sum(y == 1) / n;
        gini = 1 - (p^2 + (1 - p)^2);
    end
end

function vote = majorityVote(y)
    % majorityVote returns the majority class from the labels.
    if sum(y == 1) >= sum(y == -1)
        vote = 1;
    else
        vote = -1;
    end
end

function preds = predictTree(tree, X)
    % predictTree applies the decision tree to each row of X.
    n = size(X,1);
    preds = zeros(n,1);
    for i = 1:n
        preds(i) = traverseTree(tree, X(i,:));
    end
end

function label = traverseTree(tree, x)
    % traverseTree recursively traverses the tree for a single sample.
    if tree.isLeaf
        label = tree.prediction;
    else
        if x(tree.feature) < tree.threshold
            label = traverseTree(tree.left, x);
        else
            label = traverseTree(tree.right, x);
        end
    end
end

File: Books/A_Comprehensive_Approach_to_ML_and_AI/python/decision_tree_tutorial.py
#!/usr/bin/env python3
"""
decisionTreeTutorial.py
-------------------------
This tutorial demonstrates a decision tree classifier built entirely from scratch.
It generates a 2D dataset with two classes, recursively builds a tree using the Gini impurity,
and visualizes the decision boundary.
"""

import numpy as np
import matplotlib.pyplot as plt

np.random.seed(1)

# 1. Generate Synthetic Data
N = 100
# Class +1: centered at (2,2)
X1 = np.random.randn(N, 2) + 2
# Class -1: centered at (-2,-2)
X2 = np.random.randn(N, 2) - 2
X = np.vstack((X1, X2))
y = np.concatenate((np.ones(N), -np.ones(N)))

# 2. Build Decision Tree from Scratch

def majority_vote(y):
    return 1 if np.sum(y == 1) >= np.sum(y == -1) else -1

def gini_impurity(y):
    n = len(y)
    if n == 0:
        return 0
    p = np.sum(y == 1) / n
    return 1 - (p**2 + (1-p)**2)

def find_best_split(X, y):
    n, d = X.shape
    best_impurity = np.inf
    best_feature, best_threshold, best_split = None, None, None
    for j in range(d):
        values = np.unique(X[:, j])
        if len(values) < 2:
            continue
        thresholds = 0.5 * (values[:-1] + values[1:])
        for thresh in thresholds:
            split = X[:, j] < thresh
            if split.sum() == 0 or (~split).sum() == 0:
                continue
            impurity_left = gini_impurity(y[split])
            impurity_right = gini_impurity(y[~split])
            weighted_impurity = (split.sum()/n)*impurity_left + ((~split).sum()/n)*impurity_right
            if weighted_impurity < best_impurity:
                best_impurity = weighted_impurity
                best_feature = j
                best_threshold = thresh
                best_split = split
    return best_feature, best_threshold, best_impurity, best_split

def build_tree(X, y, depth, max_depth):
    # stopping criteria
    if (np.all(y == y[0])) or (depth >= max_depth) or (X.shape[0] < 2):
        return {'is_leaf': True, 'prediction': majority_vote(y)}
    best_feature, best_threshold, best_impurity, split = find_best_split(X, y)
    if best_feature is None:
        return {'is_leaf': True, 'prediction': majority_vote(y)}
    # If one side is empty, make leaf
    if split.sum() == 0 or (~split).sum() == 0:
        return {'is_leaf': True, 'prediction': majority_vote(y)}
    left_tree = build_tree(X[split], y[split], depth+1, max_depth)
    right_tree = build_tree(X[~split], y[~split], depth+1, max_depth)
    return {'is_leaf': False,
            'feature': best_feature,
            'threshold': best_threshold,
            'left': left_tree,
            'right': right_tree}

def predict_tree(tree, X):
    n = X.shape[0]
    preds = np.zeros(n)
    for i in range(n):
        preds[i] = traverse_tree(tree, X[i, :])
    return preds

def traverse_tree(tree, x):
    if tree['is_leaf']:
        return tree['prediction']
    if x[tree['feature']] < tree['threshold']:
        return traverse_tree(tree['left'], x)
    else:
        return traverse_tree(tree['right'], x)

max_depth = 3
tree = build_tree(X, y, depth=0, max_depth=max_depth)

# 3. Visualize Decision Boundary
x_min, x_max = X[:,0].min()-1, X[:,0].max()+1
y_min, y_max = X[:,1].min()-1, X[:,1].max()+1
xx, yy = np.meshgrid(np.linspace(x_min, x_max, 200),
                     np.linspace(y_min, y_max, 200))
grid_points = np.c_[xx.ravel(), yy.ravel()]
preds = predict_tree(tree, grid_points).reshape(xx.shape)

plt.figure(figsize=(8,6))
plt.contourf(xx, yy, preds, levels=[-np.inf, 0, np.inf], alpha=0.3, colors=['lightcoral', 'lightblue'])
plt.scatter(X[y==1,0], X[y==1,1], c='b', marker='o', label='Class +1')
plt.scatter(X[y==-1,0], X[y==-1,1], c='r', marker='o', label='Class -1')
plt.title("Decision Tree Classification")
plt.xlabel("Feature 1")
plt.ylabel("Feature 2")
plt.legend(loc='upper right')
plt.grid(True)
plt.show()
